### DISCUSSION TITLE
Vite error when importing from .server file. "Error: Server-only module referenced by client"

### DISCUSSION THREAD

**Discussion by joaoguidev** [remix-run/remix#9035] [Q&A]
Hello.

I have a file named validation.server.jsx that I use to export a schema validation function.

As the name indicates, it is a .server file with code to run exclusively server-side.

I then use the exported function from the referred file to validate data inside the action of my route.jsx.

Now with Vite, Im getting the error below just by importing the function from the validation.server.jsx file into my app/routes/admin/route.jsx.

`import { validateSchema } from "../../lib/validation.server"`

By making this import I get the error or any import from a .server file into a route.jsx file i get the error:
```
Pre-transform error: Server-only module referenced by client
    '../../lib/validation.server' imported by route 'app/routes/admin/route.jsx'
  Remix automatically removes server-code from these exports:
    `loader`, `action`, `headers`
```
It seems that it was dealt with before but im still getting this problem: [https://github.com/remix-run/remix/pull/8267](https://github.com/remix-run/remix/pull/8267)

**Comment by spacepolice10** [comment]
#8448 maybe you'll find something useful here

**Reply by joaoguidev** [reply]
It helped indeed.

The solution in my case was making the import from inside the loader or action.

instead of making the import normally on the top of my route.jsx file like this:

`import { validateSchema } from "../../lib/validation.server"`

I made the import inside the action itself like this:

```
export const action= async ({ request, context }) => {
   const { validateSchema } = await import('../../lib/validation.server');
...
```

It feels a little like a step back because methods that I import from .server file like protectRoute that I use both on loaders and actions I will need to import the same thing 2 times, 1 inside the loader and 1 inside the action on the same route.jsx file.

**Reply by kiliman** [reply]
@joaoguidev That is odd. In my `remix-vite-template` I import from `*.server` files without issues. Could it be something in your validation file?

https://github.com/kiliman/remix-vite-template/blob/8d3c4f324681097fd6aa5d31bc0426d01e8111e4/app/routes/counter.tsx#L3

**Reply by joaoguidev** [reply]
Hello Kiliman.

I noticed that different from your project, I'm using folders in my routes. 

I wonder if Vite and this style of structuring the routes might have something into it.  Before migrating to Vite I had no such issue.

So my routes look like this:
![image](https://github.com/remix-run/remix/assets/63625334/a4f2f546-7026-4aaf-8ed0-06ba8fc0e780)

Instead of this:
![image](https://github.com/remix-run/remix/assets/63625334/00842b92-db22-406c-9bba-1f5181abdc00)

Documentation [here](https://remix.run/docs/en/main/file-conventions/routes#folders-for-organization)

Another point worth mentioning is that I'm using the template for Cloudflare:  
`npx create-remix@latest --template remix-run/remix/templates/vite-cloudflare`

**Reply by kiliman** [reply]
Thanks for providing more information. I'll see if I can reproduce the problem you're having.

**Reply by toptal-dave** [reply]
> It helped indeed.
> 
> The solution in my case was making the import from inside the loader or action.
> 
> instead of making the import normally on the top of my route.jsx file like this:
> 
> `import { validateSchema } from "../../lib/validation.server"`
> 
> I made the import inside the action itself like this:
> 
> ```
> export const action= async ({ request, context }) => {
>    const { validateSchema } = await import('../../lib/validation.server');
> ...
> ```
> 
> It feels a little like a step back because methods that I import from .server file like protectRoute that I use both on loaders and actions I will need to import the same thing 2 times, 1 inside the loader and 1 inside the action on the same route.jsx file.

This worked for me as well, thanks.  However, it's definitely a step backwards because I have to manually type the import e.g. `const stripe = await import('~/lib/stripe-instance/server'}; const session = (stripe as unknown as Stripe).checkout.session.retrieve(...)`.

For what it's worth, I'm also using the folder-based routing, which is great by the way, so I would rather keep it as such.

**Reply by fjuren** [reply]
I get the same error as you @joaoguidev and I also have a nested folder structure in /routes. That said, I don't think this is the issue since the [epic stack](https://github.com/epicweb-dev/epic-stack/tree/main/app/routes) follows this structure, and it too follows imports from `*.server` files without issue.

**Comment by Marcdj-02** [comment]
I am encountering a similar problem. Yet don't know if it's related. 

```
[1] ✘ [ERROR] Server-only module referenced by client
[1] 
[1]     '~/utils/services/xxx.server' imported by route 'app/yy/zz/resume.tsx'
[1] 
[1]   Remix automatically removes server-code from these exports:
[1]     `loader`, `action`, `headers`
[1]
[1]   But other route exports in 'app/yy/zz/resume.tsx' depend on '~/utils/services/xxx.server'.
```

Even though the specified import is only used within the action function.

Everything works great when using `npm run build` and `npm start`. Yet when using `remix vite:dev` I get this error.
The error was introduced in my code base since I added the following configuration to the `vite.config.ts`:

```typescript
  optimizeDeps: {
    entries: [
      '**/routes/**/*.{ts,tsx}',
    ],
  },
```

Anyone got an idea about this?

**Reply by EnriqueCepeda** [reply]
I'm having the same problem, event when the imported function is only used in the loader

**Reply by toptal-dave** [reply]
Me too, whereas I seem to be able to use it in action exports just fine.

**Reply by agathiyans28** [reply]
Specifying all routes like `**/routes/**/*.{ts,tsx} `for custom pre-bundling in Vite isn't necessarily the best approach, especially if you have a large number of routes. This can lead to unnecessary pre-bundling of files that don't need it, increasing build times and potentially causing other inefficiencies.

- Instead of pre-bundling all routes, focus on the main entry points of your application, such as your main file (index.js, main.js, App.vue, root.tsx, etc.) and any critical dynamic imports that you know will benefit from pre-bundling.

like, 
` optimizeDeps: {
    entries: ['./app/root.tsx']
  },`

its working for me 😇

**Alternative: Use Vite's Auto-Detection**

- Vite's default behavior is usually sufficient for most projects. It automatically scans your main HTML file (or specified entries) and pre-bundles the dependencies used there. If you don't have a specific reason to customize the entries, you might not need to specify them at all.

**Reply by toptal-dave** [reply]
Not sure what happened but I was going to try the solution @agathiyans28 suggested today and when I went to replicate the build error, I couldn't.  Maybe Vite's Auto-Detection kicked in or maybe being on a different computer made it work differently but I no longer get this error.

I tried adding `optimizeDeps` and see no difference in build time nor in any of the build outputs (sizes and compression etc.).  Perhaps Vite is doing this automatically?

**Reply by agathiyans28** [reply]
Vite's auto-detection can sometimes handle dependency optimization without manual intervention, so that might explain why adding optimizeDeps didn’t change the build output. It could also be related to differences in the environment between your computers.

If everything is working fine now, that’s great. 😄

**Comment by arunaugustine** [comment]
For what it's worth:

 I had a similar problem:

` Internal server error: Server-only module referenced by client`

but it turned out, I had imported a function from the `.server.js` file but was not actually calling that function in my loader (it was commented out). When I uncommented that line, the error went away. 

I assume this is because remix determines if an import from a .server.js file is actually used in the loader function when deciding to throw this error or not.

**Reply by vritzka** [reply]
This fixed it for me.

importing from a .server file causes the error if the imported function Is NOT used in loader() or action().

**Reply by navdeep-soni** [reply]
yes it worked for me as well.
if importing from .server.js then it has to be used in loader. I was expecting a better error message for this :)

**Reply by isriram** [reply]
This does not work for me unfortunately. I still get the same error :(

```
export async function loader({ request }: LoaderFunctionArgs) {
  const { getSession } = await import("~/services/session.server");
  const { identityService } = await import("~/services/identity.server");
  let session = await getSession(request.headers.get("cookie"));
  let userSession = (await session.get(identityService.sessionKey)) as IUserSession;
  if (!userSession) {
    redirect(`/user/profile/login`);
  }
  // let userSession = {
  //   _id: "12345",
  //   email: "johnsmith@gmail.com",
  //   first_name: "John",
  //   last_name: "Smith",
  //   scope: "read_home",
  //   logged_in: "yes",
  // };
  console.log("Server-side");
  console.log(userSession);
  return userSession || null;
}
```

**Reply by navdeep-soni** [reply]
@isriram make sure you are not importing something else from xxxx.server anywhere else and not using it

also as a test return some hard coded value

**Reply by isriram** [reply]
@navdeep-soni I figured out the problem. I was using that in a component. This used to work in the classic remix compiler, however in the vite compiler it seems that .server.ts files can only be referenced from the routes.
Now everything is working as intended.

**Reply by canvasofgod** [reply]
I had the same problem, and this worked for me!

**Reply by sktlmarat** [reply]
@isriram thank you, this resolved the issue for me!

**Reply by Jeremy-RemyVentures** [reply]
This was my issue as well, have the import but commented out the method call temporarily, and the error showed up. Nice catch!

**Reply by ojolowoblue** [reply]
This solved it for me, thanks @arunaugustine.

**Comment by toptal-dave** [comment]
OK now I'm having trouble integrating Trigger.dev because their initialization, which is based on file name routing rather than folder name routing, triggers this error.  The automatically generated `app/routes/api.trigger.ts` file (whether kept that way or renamed to `app/routes/api.trigger/route.ts`) looks like this:

```ts
import { createRemixRoute } from "@trigger.dev/remix";
import { client } from "~/trigger.server";

// Remix will automatically strip files with side effects
// So you need to *export* your Job definitions like this:
export * from "~/jobs/example.server";

export const { action } = createRemixRoute(client);
```

And when I try to build I get the error:

```
x Build failed in 338ms
[commonjs--resolver] Server-only module referenced by client

    '~/trigger.server' imported by route 'app/routes/api.trigger/route.ts'

  Remix automatically removes server-code from these exports:
    `loader`, `action`, `headers`

  But other route exports in 'app/routes/api.trigger/route.ts' depend on '~/trigger.server'.

  See https://remix.run/docs/en/main/guides/vite#splitting-up-client-and-server-code

    at Object.resolveId (/home/davidrhoderick/Development/sukooni/node_modules/@remix-run/dev/dist/vite/plugin.js:964:15)
    at async PluginDriver.hookFirstAndGetPlugin (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:19792:28)
    at async resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:18466:26)
    at async ModuleLoader.resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:18869:15)
    at async Object.resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:12833:10)
    at async PluginDriver.hookFirstAndGetPlugin (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:19792:28)
    at async resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:18466:26)
    at async ModuleLoader.resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:18869:15)
    at async file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:19173:50 {
  code: 'PLUGIN_ERROR',
  plugin: 'commonjs--resolver',
  hook: 'resolveId',
```

Now, I don't think this should be an error and I think it's on the Remix side, but I'm not sure if anyone is working on this issue or if it's just not going to be fixed but I'd like to be able to integrate Remix with Trigger.dev so I'll try to see if they have any tips as well.

**Reply by toptal-dave** [reply]
> OK now I'm having trouble integrating Trigger.dev because their initialization, which is based on file name routing rather than folder name routing, triggers this error. The automatically generated `app/routes/api.trigger.ts` file (whether kept that way or renamed to `app/routes/api.trigger/route.ts`) looks like this:
> 
> ```ts
> import { createRemixRoute } from "@trigger.dev/remix";
> import { client } from "~/trigger.server";
> 
> // Remix will automatically strip files with side effects
> // So you need to *export* your Job definitions like this:
> export * from "~/jobs/example.server";
> 
> export const { action } = createRemixRoute(client);
> ```
> 
> And when I try to build I get the error:
> 
> ```
> x Build failed in 338ms
> [commonjs--resolver] Server-only module referenced by client
> 
>     '~/trigger.server' imported by route 'app/routes/api.trigger/route.ts'
> 
>   Remix automatically removes server-code from these exports:
>     `loader`, `action`, `headers`
> 
>   But other route exports in 'app/routes/api.trigger/route.ts' depend on '~/trigger.server'.
> 
>   See https://remix.run/docs/en/main/guides/vite#splitting-up-client-and-server-code
> 
>     at Object.resolveId (/home/davidrhoderick/Development/sukooni/node_modules/@remix-run/dev/dist/vite/plugin.js:964:15)
>     at async PluginDriver.hookFirstAndGetPlugin (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:19792:28)
>     at async resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:18466:26)
>     at async ModuleLoader.resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:18869:15)
>     at async Object.resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:12833:10)
>     at async PluginDriver.hookFirstAndGetPlugin (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:19792:28)
>     at async resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:18466:26)
>     at async ModuleLoader.resolveId (file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:18869:15)
>     at async file:///home/davidrhoderick/Development/sukooni/node_modules/rollup/dist/es/shared/node-entry.js:19173:50 {
>   code: 'PLUGIN_ERROR',
>   plugin: 'commonjs--resolver',
>   hook: 'resolveId',
> ```
> 
> Now, I don't think this should be an error and I think it's on the Remix side, but I'm not sure if anyone is working on this issue or if it's just not going to be fixed but I'd like to be able to integrate Remix with Trigger.dev so I'll try to see if they have any tips as well.

OK I've upgraded to Trigger.dev version 3 and this is no longer a problem.  I thought I was initializing a version 3 application automatically since I was approved into the beta but I had to upgrade manually.  When Trigger.dev is in version 3, it looks like there are no Remix-based web-hook URLs, so this shouldn't be a problem.

**Reply by akomm** [reply]
The code in your above post has a side effect, so the action code is not removed from client bundle.
https://remix.run/docs/en/main/guides/constraints#no-module-side-effects

**Comment by Knorway** [comment]
it also happens when you re-export route relevant modules imported from other file.
```
├── _index
│   ├── +loader.tsx
│   └── route.tsx
```
```typescript
`+loader.tsx`
import { db, post } from '~/server/driver/db.server';

export const loader: LoaderFunction = async () => {
	const result = await db.select().from(post).all();
	return result;
};
...clientLoader, meta, etc
```
```typescript
`route.tsx`
...
export { loader, clientLoader, meta, shouldRevalidate } from './+loader';
```

I splitted +loader.file into two, one that exports `loader`, `headers`, `action` and another for everything else and it seems working.
```typescript
export { clientLoader, meta, shouldRevalidate } from './+meta';
export { action, headers, loader } from './+server';
export default Route;
```

**Comment by akomm** [comment]
This issue arises, when you have *.server files / folder used outside route modules, because remix does not remove *.server imports. It removes just imports from route modules which are only used within loader and action. If it happens to be a *.server import from the route module file, then everything is fine. But when the *.server import is hidden further down the import-chain, remix isn't going to remove it and so the server-only check will fail.

Additionally, pure `type` imports in client code from a *.server file used *outside* action and loader will still trigger the error too. Because the checker seems to simply analyse present imports. Might also differ in behavior based on `import type {X}` or `import {type X}` used - the later one would typically leave `import {} from "..."` (expect behavior) in the end file. But AFAIK it did not matter.

**Reply by mbrowne** [reply]
That's too bad that it doesn't analyze files further down the import chain. But I discovered a workaround—using a dynamic import:

```js
import type { LoaderFunctionArgs } from '@remix-run/node'

export async function loader(args: LoaderFunctionArgs) {
    // do a dynamic import to work around remix's lack of support for re-exporting
    // loaders that depend on server-only modules
    const importedLoader = (
        await import(
            '~/routes/my-route/my-loader.server'
        )
    ).loader

    return importedLoader(args)
}
```

Another workaround would be to use https://github.com/pcattori/vite-env-only

**Reply by akomm** [reply]
The package `vite-env-only`, by its description, only provides errors when you import the "wrong" files from the "wrong" location. You already receive error. Your problem is though the expectation of those exports being pruned.

The behavior makes sense:

imagine you import a file `foo.ts` outside loader that imports a `bar.server.ts` file.  If `foo.ts` depends on `bar.server.ts`, simply removing the import would break your code. Knowing and understanding dependencies is costly and error prone. Remix does it relatively simply, by removing all imports from route file that are only referenced in action and loader.

Though I think there *might* be a thing that could be improved, because it seems to me its not the case currently:
Either the `server-only` check runs *before* tree-shaking, or its run after, but when you import *just* types the "wrong" way:
```ts
// bad: notice type is within {}
import {type Foo} from "./foo" // -> leaves import {} from "..."
// good: notice type outside {}
import type {Foo} from "./foo"
```
it traverses the empty import and hits some potential `.server` code.

In case its run before tree-shaking, maybe it can be changed to run after.
In case its the type import issue, maybe the build can be improved to remove empty name exports, prior to server-only check.

The later would assume that when you don't import just as a side-effect `import "..."`, the left over `import {} from "..."` is never allowed to have a side-effect. This assumption may be false and cause issues, where dev works and prod build fails, because something worked in dev by having an side-effect run (unintentionally) and this side-effect isn't going to run in prod later.

**Reply by cesarh1094** [reply]
I was experiencing the a similar issue as the OP with regards to importing a _type_ from a server-only file to non-server-only file.

I wonder if there's something in the Vite or Remix/react-router docs than can provide further clarity why you need to make sure the `type` keyword is outside of the "things" you're importing. I would consider that this to be important distinction especially with how react-router tree-shakes server/client code for the respective bundles.

..either way. Thanks for providing the solution here! 🙏🏾

**Reply by akomm** [reply]
Has nothing to do with remix or react-router.  

Types do not exist at runtime, so `import {type Foo} from "./foo.ts"` will end up being `import {} from "./foo.ts"`. The module will be loaded and any potential side-effects executed. A different behavior would mean a behavior divergence and potential bugs between development and production.

In order to depict explicitly, that you only want to load types (and not load the module, hence execute potential side-effects) is to use a pure type import `import type {Foo} from "./foo.ts"`. Example:

```ts
// bar.ts
export type Bar = {}

console.log('bar side-effect')

// foo.ts
export type Foo = {}

console.log('foo side-effect')

// test.ts
import {type Bar} from "./bar"
import type {Foo} from "./foo"

function test(foo: Foo, bar: Bar) {}

console.log("end")

```

Output: 
```
bar side-effect
end
```

**Reply by mbrowne** [reply]
@akomm that's true about type imports being a way to avoid any effect on run-time, but the issue in the OP affects real imports as well.

**Reply by akomm** [reply]
I was replying to a comment, not OP?

**Comment by Pinwheeler** [comment]
okay so if we're not supposed to use actions outside of routes, how do I do something like... have a signout button in a header that's present everywhere?

It seems heavy-handed that Remix should dictate that my header must be part of my dashboard component and can't be a separate file...

**Reply by AntonioAngelino** [reply]
Create a /signout page that performs the action. Redirect the user there when he clicks the btn.

**Reply by herkulano** [reply]
You can also use a resource route for the signout with an action only to get the POST request: https://reactrouter.com/how-to/resource-routes

**Reply by Pinwheeler** [reply]
thanks for the thoughtful suggestions, both :)

**Comment by seanmoghadam** [comment]
We faced this issue during the migration from Remix v2 to RR7. We were using **remix-flat-routes** that are not configured correctly if you only follow the migration guide.
After applying the recommended configuration to the routes.ts file as described [here](https://github.com/kiliman/remix-flat-routes?tab=readme-ov-file#-react-router-v7-support):

```import { remixRoutesOptionAdapter } from "@react-router/remix-routes-option-adapter";
import { flatRoutes } from "remix-flat-routes";

export default remixRoutesOptionAdapter((defineRoutes) => {
  return flatRoutes("routes", defineRoutes, {
    ignoredRouteFiles: ["**/.*"], // Ignore dot files (like .DS_Store)
    //appDir: 'app',
    //routeDir: 'routes',
    //basePath: '/',
    //paramPrefixChar: '$',
    //nestedDirectoryChar: '+',
    //routeRegex: /((\${nestedDirectoryChar}[\/\\][^\/\\:?*]+)|[\/\\]((index|route|layout|page)|(_[^\/\\:?*]+)|([^\/\\:?*]+\.route)))\.(ts|tsx|js|jsx|md|mdx)$$/,
  });
});
```

Vite could pick up the routes correctly and strip away the server modules during the build and local "dev" process. Hope this helps!

**Comment by MrTechHunter** [comment]
## 🐛 Error: "Server-only module referenced by client" (Vite + Remix)

✅ Solution found — scroll down for step-by-step fix

We ran into the following Vite error when trying to use a `.server.ts` file inside a client-side hook in Remix:

```
Error: Server-only module referenced by client
```

---

## 💡 Context

**File structure:**

```
app/
├── hooks/
│   └── useSendProducts.ts        ← Client-side hook
├── utils/
│   └── example.server.ts         ← Server-only code
├── routes/
│   └── example-page.tsx          ← Uses loader() and renders data
```

In `useSendProducts.ts`, we tried:

```ts
import { callAnAPI } from "../utils/example.server"; // ❌ Causes the error
```

This breaks because client-side code can’t directly import server-only modules.

---

## ✅ Solution: Use an API route to bridge client ↔ server

We resolved it by moving the server logic into a server-side action route and calling it from the hook using `useFetcher`.

---

## 🔧 Step-by-Step Fix

### 1. Create a server-side action route

**`app/routes/api.send-products.tsx`**

```ts
import type { ActionFunction } from "@remix-run/node";
import { json } from "@remix-run/node";

export const action: ActionFunction = async ({ request, context }) => {
  const { callAnAPI } = await import("../utils/example.server");
  const { session } = context;
  const shop = session?.shop;

  const formData = await request.formData();
  const chunk = JSON.parse(formData.get("chunk") as string);

  try {
    const response = await callAnAPI(shop, chunk);
    if (response.ok) {
      return json({ success: true, failedProducts: [] });
    }
    return json({ success: false, failedProducts: chunk });
  } catch (error: any) {
    console.error("Server error:", error);
    return json({ success: false, failedProducts: chunk, error: error.message }, { status: 500 });
  }
};
```

---

### 2. Call the server route from the hook

**`app/hooks/useSendProducts.ts`**

```ts
import { useFetcher } from "@remix-run/react";

export function useSendProducts() {
  const fetcher = useFetcher();

  const sendProducts = async (chunk: Product[]): Promise<{ success: boolean; failedProducts: Product[] }> => {
    try {
      const formData = new FormData();
      formData.append("chunk", JSON.stringify(chunk));

      fetcher.submit(formData, {
        method: "POST",
        action: "/api/send-products",
      });

      return new Promise((resolve) => {
        const checkFetcher = setInterval(() => {
          if (fetcher.state === "idle" && fetcher.data) {
            clearInterval(checkFetcher);
            const { success, failedProducts, error } = fetcher.data;
            if (error) console.error("Fetcher error:", error);
            resolve({ success, failedProducts: failedProducts || chunk });
          }
        }, 100);
      });
    } catch (error) {
      console.error("Error sending products:", error);
      return { success: false, failedProducts: chunk };
    }
  };

  return { sendProducts };
}
```

---

### 3. Use server-only logic in the loader (safe)

**`app/routes/example-page.tsx`**

```ts
import { json } from "@remix-run/node";
import type { LoaderFunctionArgs } from "@remix-run/node";

export async function loader({ request }: LoaderFunctionArgs) {
  const { getProducts } = await import("../utils/example.server");
  const { session } = await authenticate.admin(request);
  const shop = session.shop;

  const result = await getProducts(shop);
  return json({ result });
}
```

**In your component:**

```tsx
export default function ExamplePage() {
  const { result } = useLoaderData<typeof loader>();
  return <pre>{JSON.stringify(result, null, 2)}</pre>;
}
```

---

## ✅ Outcome

* No more Vite errors
* `.server.ts` files only run on the server
* Client communicates via `useFetcher` + route-based actions

**Comment by MrTechHunter** [comment]
Hi @kiliman,

I’ve shared a detailed answer in the discussion [#9035](https://github.com/remix-run/remix/discussions/9035#discussioncomment-13022894) that I believe could help others facing the same issue. Would it be possible to pin or mark my comment as the answer to make it more visible to the community? I’d appreciate your consideration!

Thanks,

Mahdi Hamldar

**Reply by akomm** [reply]
The answer feels totally detached from current conversation state. It does not answer why error happens (has been answered above many times). It shoots with a canon using weird `setInterval` and Promise hacks around the fetcher for a trivial problem, if you understand why it happens. And also the naming feels mixed/blend  together, like why are products called "chunk". Its typical, weird mixing I have seen often in AI generated code... Whats also typically blend is that it starts as step-by-step and "example" approach, but then blends in code that is not example but has a lot specifics, like the context out of nowhere. And what is "shop" from context? Why is it relevant to the problem?

Your example even needlessly uses async import, because at this position a regular import would have been pruned by remix and the error would not haven appeared either. The issue was the usage of the .server file in client code outside action/loader.

**Comment by danpen-S** [comment]
All of these solutions seem backwards. The entire point of Remix is to write full-stack applications. Loaders, actions, and the rendered component/page are all in the same file. Loaders and actions should be able to import .server files

Observations:

1. This is how it was done before Vite was a thing.
2. Even when using Remix+Vite, `npx remix dev` allows these imports. (`npm run build` does not)
3. In my case, the .server imports aren't even being used by the component/page, only by the loader and/or action

**Reply by akomm** [reply]
Loader and action function *are* able to import .server files. Remix will remove this imports if used only in loader or action. 

Any time I had this apparent situation myself, it turned out some non-server file, within which a .server import is hidden somewhere deep triggered it. Maybe even to the same file you import and use *only* in action and loader. So you see the file you apparently *only* use in loader and action, but the error on that file was not triggered via the route module import, but via some other import that *also* end up pulling the same .server file.
